bank2ynab/plugins/parse_from_memo.py

```diff
@@ -0,0 +1,138 @@
+"""
+Plugin that uses information from the memo field to fill in other fields
+"""
+
+# Usage:
+# Ensure the relevant bank config has a Memo Parser config attribute
+# The attribute should contain a list of regular expressions, run from top to bottom
+# with named groups, used to override fields in the CSV.
+#   Memo Parser =
+#       (?<payee>\w+) (?<memo>.*)? AT (?<time>\d{2}\.\d{2})
+#       PURCHASER: (?<purchaser>\w+) AT (?<memo>.*)?
+# Valid groups are: payee, memo, date, time, purchaser
+# At that point, you can reference the plugin in conf files like this:
+#   Plugin = parse_from_memo
+
+import re
+import logging
+from datetime import datetime
+
+from bank_handler import BankHandler, get_output_path
+from dataframe_handler import read_csv
+from transactionfile_reader import detect_encoding
+
+logging.basicConfig(format="%(levelname)s: %(message)s", level=logging.DEBUG)
+
+class ParseFromMemo(BankHandler):
+    def __init__(self, config_object):
+        """
+        :param config_object: a dictionary of conf parameters
+        """
+        super(ParseFromMemo, self).__init__(config_object)
+        self.name = "ParseFromMemo"
+
+        # Parsers from the Config, skipping blank rows
+        memo_parsers = list(filter(len, self.config_dict.get("plugin_args", [])))
+        if len(memo_parsers) <= 0:
+            raise AttributeError(
+                "The plugin arguments option contained no regular expresions to use for parsing"
+            )
+        self.parsers_to_try = list(
+            map(re.compile, memo_parsers)  # convert to regex
+        )
+
+    def _preprocess_file(self, file_path: str, plugin_args: list) -> str:
+        """
+        Override the memo, date and payee columns with information read from the memo field
+        """
+
+        # Read in the old CSV
+        src_encod = detect_encoding(file_path)
+        df = read_csv(
+            file_path=file_path,
+            delim=self.config_dict.get("input_delimiter", ','),
+            header_rows=0,
+            footer_rows=0,
+            encod=src_encod,
+        )
+
+        # Apply our adjustment
+        new_df = df.apply(self._parse_from_memo, axis="columns")
+
+        # Generate output path
+        new_path = get_output_path(
+            input_path=file_path,
+            prefix=f"memo_parsed_csv_{self.config_dict['bank_name']}_",
+            ext=".csv",
+        )
+
+        # Write the dataframe to output file
+        new_df.to_csv(new_path,
```

Comment:
Having to read from the CSV, to write into a CSV and re-load is a less than ideal plugin interface. I originally built against master and manipulated the data _after_ the BankHandler had read in the CSV; however it looks like this is the best way with the "new" plugin API.

Reply:
I know, it's a bit clunky, but the logic was that a plugin should exist to get the file into a "base" state where all normal operations can then be applied on it. It just seemed like the most universal approach.